### PR TITLE
Expression matching API

### PR DESCRIPTION
**Pull request by tlively**
Provides an easily extensible layered API for matching expression patterns and
extracting their components. The low-level API provides modular building blocks
for creating matchers for any data type and the high-level API provides a
succinct and flexible interface for matching expressions and extracting useful
information from them.

Matchers are currently provided for Const, Unary, Binary, and Select
instructions. Adding a matcher for a new type of expression is straightforward
enough that I expect to add them as they become useful as part of other changes.

cc @MaxGraey

### CODE REVIEW

**Comment by tlively** [comment]
There won't be any more full rewrites of this, I promise ;)

File: src/passes/OptimizeInstructions.cpp
Comment (MaxGraey) [inline src/passes/OptimizeInstructions.cpp:1329]: Hmm, it should check `value` with zero I guess:
```cpp
if (value == 0.0) {
  // x - (-0.0)   ==>   x + 0.0
  if (std::signbit(value)) {
    ...
  } else {
    ...
  }
}
```

File: src/passes/OptimizeInstructions.cpp
Comment (MaxGraey) [inline src/passes/OptimizeInstructions.cpp:1337]: the same

File: src/passes/OptimizeInstructions.cpp
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:1329]: Or you could generalize `x - (-C)` ==> `x + C`. In this case:
```cpp
if (matches(curr, binary(Abstract::Sub, any(), fval(&value)))) {
  // x - (-C)   ==>   x + C
  if (std::signbit(value)) {
    curr->op = Abstract::getBinary(type, Abstract::Add);
    right->value = right->value.neg();
    return curr;
  } else {
    if (value == 0.0) {
      // x - 0.0   ==>   x
      return curr->left;
    }
  }
}
```

Comment (tlively) [reply src/passes/OptimizeInstructions.cpp:1329]: Good catch, thanks! Is that transformation of `X - (-C)` always correct?

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:1329]: Yes, even with non-canonical `-nan` witch become canonical `nan`. Hope fuzz tests also happy with this

File: src/passes/OptimizeInstructions.cpp
Comment (MaxGraey) [inline src/passes/OptimizeInstructions.cpp:1337]:
```suggestion
          std::signbit(value) && value == 0.0) {
```

File: src/passes/OptimizeInstructions.cpp
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:197]: If we think `commutable` is a better name than `canReorder`, let's add a TODO to rename the latter? (as it seems odd to have it call a method with a different name here)

However I'm not sure I prefer it, as `commutable` may have a wider meaning in general, can't it? (not just ordering)

File: src/passes/OptimizeInstructions.cpp
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:229]: may as well hoist and unify these `x, y` etc. declarations?

File: src/passes/OptimizeInstructions.cpp
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:230]: we could avoid the `&` here by using `Expression*&` in the API, which would be shorter? or is the idea this is more explicit?

Comment (tlively) [reply src/passes/OptimizeInstructions.cpp:230]: Yeah, I actually like having `&` as a sigil to differentiate binders for extracted values. It also allows us to provide more functionality, e.g. having both `i32(int32_t)`, which matches exact values, and `i32(int32_t*)` which matches and extracts any value.

File: src/passes/OptimizeInstructions.cpp
Comment (tlively) [reply src/passes/OptimizeInstructions.cpp:229]: I think it's actually better for readability to declare any necessary binder variables at the location where the pattern is used. That way, the reader only needs to read a small, contiguous block of code to fully understand what transformation is being applied. It also allows anyone adding a new pattern to use `x` with whatever type makes sense for that pattern.

File: src/passes/OptimizeInstructions.cpp
Comment (tlively) [reply src/passes/OptimizeInstructions.cpp]: Even though it's small, let's leave that generalization to a follow-up PR. This PR is already large and complex, so it will be helpful to avoid functional changes in OptimizeInstructions.

File: src/passes/OptimizeInstructions.cpp
Comment (tlively) [reply src/passes/OptimizeInstructions.cpp:197]: I guess normally an operation is commutable, but here I am describing the operands as commutable. I'll rename this to `canReorder`.

Comment (tlively) [reply src/passes/OptimizeInstructions.cpp]: Oh nice, it's even the same number of characters 😄

**Review by aheejin — COMMENTED** [review]
Sorry for the late reply! While I haven't done reading this yet and I'm not a template metaprograming expert, I'm curious what the main advantages of this version over #3122 are, because the user side mostly looks the same.

**Comment by tlively** [comment]
> Sorry for the late reply! While I haven't done reading this yet and I'm not a template metaprograming expert, I'm curious what the main advantages of this version over #3122 are, because the user side mostly looks the same.

The biggest user-facing advantage is that the constant matchers can now extract int32_t, int64_t, float, and double values from constants directly. This is made possibly by generalizing the matching infrastructure to work on `Literal`s and primitive types in addition to expressoins. Another advantage is that it requires much less duplicated logic to add new matchers and it is easier to extend matchers. For example, if I wanted to add a new kind of `unary` matcher that extracted the `UnaryOp`, I would just have to increase the `NumComponents` value for `UnaryKind<UnaryK>` and add a new specialization of `GetComponent` that returns the op. After that, I could use the standard `Internal::Exact` and `Internal::Any` matchers on the op as well.

**Comment by tlively** [comment]
I'll fix merge conflicts after initial review because otherwise I expect there to be a continuous train of merge conflicts as we continue to land things in OptimizeInstructions.

File: src/ir/match.h
Comment (tlively) [reply src/ir/match.h:109]: Stale comment! This should be `matched_t`.

File: src/passes/OptimizeInstructions.cpp
Comment (tlively) [reply src/passes/OptimizeInstructions.cpp:312]: Actually, I think it would be best to extract these select optimizations into their own function. Like your suggestion, this will also have the effect of having the top-level match compare with a `Select*`, which might save a dynCast after inlining. I definitely want to continue matching from the root of the expression, though, to make it easy to see what is being matched with no context.

**Review by aheejin — COMMENTED** [review]
Looks very thorough, modular and composable to me. I'm not a template metaprogramming expert, so while I can understand what this code does, I don't have a lot of insights on improving technicalities, but to me it looks very nice.

One slight concern was that while this is strictly better on modularity and composability that the previous version, metaprogramming stuff can make people little more time to understand and add things. But maybe with a good tutorial it should be fine.

I have some questions below, mostly minor.

File: src/ir/match.h
Comment (tlively) [reply src/ir/match.h:295]: Yeah, I don't think I was very consistent about that. Using references for some things will be important if I go back and add building back in at some point, but even then we wouldn't need references here. I'll go in a remove all the unneeded references for now.

File: src/ir/match.h
Comment (tlively) [reply src/ir/match.h:330]: No good reason, will remove.

File: src/ir/match.h
Comment (tlively) [reply src/ir/match.h:358]: Right. I'll add a line of documentation explicitly stating that the value of bound variables is not defined if a match fails.

File: src/ir/match.h
Comment (tlively) [reply src/ir/match.h:395]: Will make this a pass by value.

File: src/ir/match.h
Comment (tlively) [reply src/ir/match.h:573]: Since `0` can be implicitly converted to any integral type or any pointer type pointer type, there is an overload resolution ambiguity when there is both an overload taking a pointer type and an overload taking a integral type. This ambiguity is solved by adding an overload for `int`, which is the original type of the literal `0`.

I will elaborate in the comment.

File: src/ir/match.h
Comment (tlively) [reply src/ir/match.h:540]: Possibly, although I would hope this all gets inlined away anyhow. It would also have the effect of encouraging matchers to be built all in one expression as intended, rather than built from the previously initialized parts. I will make this change.

File: src/ir/match.h
Comment (tlively) [reply src/ir/match.h:450]: > Does this only support i32? Or we only have usecases for i32 and this can be extended to support other types later?

This is able to match i32, i64, f32, or f64, but its data is always an int32_t because that makes it possible to match both ints and floats. The user-facing wrapper function for this matcher is `constant`, and if the user wants to match a more restricted set of types, they can use `i32`, `f32`, `ival`, etc.

> Is that because we want to bypass the type checking?

Yes, exactly. Most of the literal matchers are able to extract a value of a known type, which can then be matched with `Any` or `Exact`. But `NumberLitKind` doesn't know what kind of value it would have to extract and there's no way to additional checking in `GetComponent`, so there's no way to use a submatcher and make it type check. Since we don't have a statically known type for a subcomponent, `NumberLitKind` has to do all its work in `MatchSelf`.

File: src/ir/match.h
Comment (tlively) [reply src/ir/match.h:735]: `constant` supports all of `i32`, `i64`, `f32`, and `f64`, but that means it only matches integer float values like `1.0` or `2.0`. Since it matches only integer floats, we could in theory allow all matched values to be bound to an int32_t, but that might give readers the mistaken impression that only `i32` and `i64` constants can be matched. Users can extract the `Literal`, though, which gives them access to both the value (if they didn't already specify it) and its type.

File: src/passes/OptimizeInstructions.cpp
Comment (tlively) [reply src/passes/OptimizeInstructions.cpp:196]: Yeah, I was actually thinking about adding a custom matcher for this so we could write things like `pure(&x)` inline in match patterns.

File: src/passes/OptimizeInstructions.cpp
Comment (kripken) [reply src/passes/OptimizeInstructions.cpp]: Wouldn't the name of the variable being "select" be enough to see what is being matched, even if the actual match was a little higher up?

Comment (kripken) [reply src/passes/OptimizeInstructions.cpp]: (I don't care much, but this does seem like it would be shorter and more efficient the way I'm suggesting, unless I'm missing something.)

**Review by kripken — APPROVED** [review]
This looks good enough to me for landing. I'd like to carefully see that each changed optimization here hasn't been altered, but it's pointless to do that while the PR is still changing (no way to mark which lines I've already read). I can do that after it lands I think, worst case we'll have a followup.

Other comments about select and other improvements can be deferred as well.

Very nice work!

**Comment by tlively** [comment]
Thanks, @kripken! Can I get an explicit signoff on just the last commit I pushed as well? I'm not sure you saw it since I pushed around the same time as your approval comment, and it's related to one of your previous comments.

**Review by aheejin — APPROVED** [review]
Haven't checked the equivalence of every single translation in OptimizeInstructions.cpp, but the whole infrastructure looks great to me! It was a fun and learning read for me too.

**Comment by kripken** [comment]
Oh, since all lgtms are in, are no further changes expected? Then give me an hour and I'll do a final pass on all optimizations here!

File: src/passes/OptimizeInstructions.cpp
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:837]: wait, does `i32()` not just check for any `i32` but also a non-zero one?

File: src/passes/OptimizeInstructions.cpp
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:1353]: this match is kind of wasteful, as we've matched most of it before... but maybe the optimizer figures it out? clarity-wise it seems a wash, so seems ok as it is.

**Review by kripken — COMMENTED** [review]
optimizations lgtm except for that one question on `i32()`...?

Also running a quick fuzzing, e.g.
```wat
(module
 (type $i64_=>_none (func (param i64)))
 (func $0 (param $0 i64)
  (if
   (i32.const 0)
   (block
    (drop
     (f64x2.sub
      (v128.const i32x4 0x00000001 0x00000000 0x00000000 0x00000000)
      (v128.const i32x4 0xcf800000 0x77745d60 0xffffd657 0x46107800)
     )
    )
    (unreachable)
   )
   (nop)
  )
 )
)
```
when run with `-O1` it crashes with
```
v128 not implemented yet
UNREACHABLE executed at /home/azakai/Dev/binaryen/src/ir/abstract.h:267!
```
Seems this reaches new places in `abstract.h`, why is that?

File: src/passes/OptimizeInstructions.cpp
Comment (tlively) [reply src/passes/OptimizeInstructions.cpp:837]: No, it's just non-zero here because zeroes would have matched above. I'll clarify that in the comment and also add an assert.

**Comment by tlively** [comment]
> Seems this reaches new places in `abstract.h`, why is that?

This is because the `AbstractBinaryMatcher` and `AbstractUnaryMatcher` call `Abstract::getBinary` and `Abstract::getUnary` in `MatchSelf` before the submatchers verify that the subexpressions have reasonable (e.g. not v128) types. I will fix this by removing the assertions about v128 in abstract.h. Handling vectors in abstract.h is not something we're going to do anytime soon.

File: src/passes/OptimizeInstructions.cpp
Comment (tlively) [reply src/passes/OptimizeInstructions.cpp:1353]: I hope the optimizer figures this out, but I'm not sure. I'll do another performance check before landing this.

**Comment by tlively** [comment]
Whatever this test failure is, it doesn't reproduce locally :'(

**Comment by kripken** [comment]
Fixes commit lgtm!

About the error, I'd try to build with both clang and gcc locally. It might be the ordering issue, stuff like this in the test
```cpp
Expression* condOne = builder.makeSelect(one(), zero(), zero());
```
Those calls are in undetermined order. That shouldn't matter on this line, but maybe it does somewhere else?

**Comment by tlively** [comment]
Aha, it's a semantic merge conflict with the wasm64 change. Will back out my debugging changes and merge this PR once that's sorted out.

**Comment by tlively** [comment]
Ok, so I measured the performance of OptimizeInstructions before and after this PR. Averaged over five runs on a 1.3MB binary:

 - With this PR: 0.091s (down from approx. 0.11s before the inlining commit)
 - Before: 0.074s

So there is about a 23% regression so far. This pass is still pretty fast, so I'm not too worried about this. but we definitely shouldn't do anything like go around turning every `dynCast` into a `matches` call. As we port patterns, we should keep an eye on the regressions, and at some point it might be worthwhile to investigate the regression more and fix it.

**Comment by MaxGraey** [comment]
I'm just wondering what if use `__attribute__((always_inline))` for Clang/GCC and `__forceinline` for MSVC instead usual `inline`? Like in [LLVM's DWARF part](https://github.com/WebAssembly/binaryen/blob/4d28d3f32e7f213e300b24bc61c3f0ac9d6e1ab6/third_party/llvm-project/include/llvm/Support/Compiler.h#L230)

**Comment by tlively** [comment]
Definitely worth experimenting with! I'd have no problem landing that if it had a measurable performance benefit.

**Pull request merged**
